### PR TITLE
Blastem - Mega Drive emulator

### PR DESCRIPTION
**Pull request by hadess**
Even includes screenshots

### CODE REVIEW

**Comment by hadess** [comment]
CC: @mikepavone

File: com.retrodev.blastem.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  <description>
+    <p>
+      BlastEm is an open source, highly accurate emulator for the Genesis/Megadrive
+      that runs on modest hardware. It also supports playing back Sega Master System games.
+    </p>
+    <p>
+      Blastem has support for savestates, Mega/Sega Mouse, XBAND and Saturn keyboard emulation.
+    </p>
+  </description>
+  <url type="homepage">https://www.retrodev.com/blastem/</url>
```
Comment (bilelmoussaoui) [inline com.retrodev.blastem.appdata.xml]: OARS tag is missing.

File: com.retrodev.blastem.json [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        /* Play sounds */
+        "--socket=pulseaudio",
+        /* Joystick and GPU access */
+        "--device=all",
+        /* Get access to the files */
+        "--filesystem=home:ro",
+        "--filesystem=xdg-run/gvfs:ro"
+    ],
+    "build-options" : {
+        "cflags" : "-O2 -g",
```
Comment (bilelmoussaoui) [inline com.retrodev.blastem.json]: No need for those build options, they are already the default ones.

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by hadess** [comment]
As expected:
```
builddir/files/share/appdata/com.retrodev.blastem.appdata.xml: FAILED:
• url-not-found         : <screenshot> failed to download (HTTP 404: Not Found) [https://github.com/flathub/com.retrodev.blastem/raw/master/screenshots/blastem-1.png]
• url-not-found         : <screenshot> failed to download (HTTP 404: Not Found) [https://github.com/flathub/com.retrodev.blastem/raw/master/screenshots/blastem-2.png]
• url-not-found         : <screenshot> failed to download (HTTP 404: Not Found) [https://github.com/flathub/com.retrodev.blastem/raw/master/screenshots/blastem-3.png]
```

I don't think there's anything to fix here. For what it's worth, this is a recurring problem when there's no screenshots that are already hosted somewhere else.

File: com.retrodev.blastem.appdata.xml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  <screenshots>
+    <screenshot type="default" width="640" height="517">https://github.com/flathub/com.retrodev.blastem/raw/master/screenshots/blastem-1.png</screenshot>
```
Comment (bilelmoussaoui) [inline com.retrodev.blastem.appdata.xml]: @hadess well, they are not hosted anywhere. The URL points to a repository that doesn't exist yet. I would suggest using your current fork for the screenshots and update those with the new URL once the repo is created.

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by nedrichards** [comment]
This all seems great. I assume based on the webpage that the author is happy just releasing binaries themselves, but we still need to reach out and let them know that they can have commit access if they want (and if not, to add it to their list of 3rd party packagers I suppose). Will add as soon as that's done.

**Comment by hadess** [comment]
> This all seems great. I assume based on the webpage that the author is happy just releasing binaries themselves, but we still need to reach out and let them know that they can have commit access if they want (and if not, to add it to their list of 3rd party packagers I suppose). Will add as soon as that's done.

The upstream was CC:ed in the 2nd comment

File: 0001-Add-support-for-Flatpak-config-data-dir-variables.patch [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
++	}
+ 	if (!confdir) {
+ 		char const *base = get_home_dir();
+ 		if (base) {
+@@ -1000,6 +1003,9 @@ char const *get_config_dir()
+ char const *get_userdata_dir()
+ {
+ 	static char* savedir;
++	if (!savedir) {
++		savedir = getenv("XDG_DATA_HOME");
```
Comment (TingPing) [inline 0001-Add-support-for-Flatpak-config-data-dir-variables.patch]: Missing context from diff, but probably should be in a subdir?

File: com.retrodev.blastem.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            "name" : "blastem",
+            "buildsystem" : "simple",
+            "build-commands" : [
+                "DATA_PATH=/app/share/blastem/ CONFIG_PATH=/app/etc/ make blastem",
+                "install -m0755 -D blastem /app/bin/blastem.bin",
+                "install -m0755 -D blastem.sh /app/bin/blastem",
+                "install -d /app/share/blastem/",
+                "cp -r shaders/ images/ gamecontrollerdb.txt rom.db /app/share/blastem/",
+                "install -m0644 -D default.cfg /app/etc/default.cfg",
+                "install -m0644 -D com.retrodev.blastem.png /app/share/icons/com.retrodev.blastem.png",
```
Comment (TingPing) [inline com.retrodev.blastem.json]: Incorrect path?

File: com.retrodev.blastem.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        "--filesystem=xdg-run/gvfs:ro"
+    ],
+    "modules" : [
+        "shared-modules/glu/glu-9.0.0.json",
+        "shared-modules/glew/glew.json",
+        {
+            "name" : "blastem",
+            "buildsystem" : "simple",
+            "build-commands" : [
+                "DATA_PATH=/app/share/blastem/ CONFIG_PATH=/app/etc/ make blastem",
```
Comment (TingPing) [inline com.retrodev.blastem.json]: `-j $FLATPAK_BUILDER_N_JOBS` (or stop using `simple` buildsystem)

File: com.retrodev.blastem.json (hunk 5 above)
Comment (hadess) [reply com.retrodev.blastem.json]: No, it's correct and exported. It's a path for icons that don't have multiple sizes, don't have sizes that correspond to the set icon sizes, or don't have scalable versions. This is also the path I use for the icons in Flash player and Teo.

**Comment by LeandroStanger** [comment]
bot, build com.retrodev.blastem

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.retrodev.blastem

**Pull request closed without merging**